### PR TITLE
Complete DVC pipeline setup and start Docker setup

### PR DESCRIPTION
**Pull request by agalmazan**
# 📝 Pull Request

## 🚀 Descripción
Este PR implementa un **pipeline completo de ML con DVC** que automatiza el flujo de trabajo de ciencia de datos desde el preprocesamiento de datos hasta la generación de predicciones. El pipeline garantiza reproducibilidad y versionado de todos los artefactos mediante DVC + AWS S3.

## 🛠 Tipo de Cambio
- [x] ✨ feat: nueva funcionalidad o script
- [x] 🔧 chore: tarea de mantenimiento
- [x] 📚 docs: actualización de documentación

### 🎯 Pipeline Principal
- `dvc.yaml` - Definición completa del pipeline con 5 etapas
- `params.yaml` - Configuración centralizada de hiperparámetros
- `dvc.lock` - Bloqueo de versiones para reproducibilidad

### 🔧 Scripts del Pipeline
- `Equipo54_MLOps/src/data/make_dataset.py` - Limpieza y preprocesamiento de datos
- `Equipo54_MLOps/src/features/build_features.py` - Ingeniería de características y encoding
- `Equipo54_MLOps/src/models/train_model.py` - Entrenamiento del modelo Decision Tree
- `Equipo54_MLOps/src/models/evaluate_model.py` - Evaluación y métricas del modelo
- `Equipo54_MLOps/src/models/predict_model.py` - Generación de predicciones

### 📊 Artefactos Generados
- `Equipo54_MLOps/models/decision_tree_model.pkl` - Modelo entrenado
- `Equipo54_MLOps/models/label_encoders.pkl` - Encoders para características categóricas
- `Equipo54_MLOps/reports/metrics/metrics.json` - Métricas de evaluación
- `Equipo54_MLOps/reports/predictions.csv` - Predicciones generadas

### 📚 Documentación
- `README_DVC.md` - Guía completa de uso de DVC + AWS S3
- `.gitignore` - Configuración para archivos DVC y datos
- `.dvc/config` - Configuración de remote S3

### 1. **process_data** 
```yaml
# Limpieza y preprocesamiento inicial
Input: student_entry_performance_modified.csv
Output: student_clean.csv
```

### 2. **build_features**
```yaml
# Ingeniería de características y encoding
Input: student_clean.csv  
Output: student_features.csv + label_encoders.pkl
```

### 3. **train**
```yaml
# Entrenamiento del modelo Decision Tree
Input: student_features.csv
Output: decision_tree_model.pkl + train_test_split.pkl
Params: Hiperparámetros desde params.yaml
```

### 4. **evaluate**
```yaml
# Evaluación del modelo y métricas
Input: Modelo + splits + encoders
Output: metrics.json + metrics.pkl
```

### 5. **predict**
```yaml
# Generación de predicciones
Input: Modelo + features
Output: predictions.csv
```

### 1. **Configuración Inicial**
```bash
# Clonar y configurar DVC
git clone <repo>
cd <project>
dvc init
dvc remote add -d myremote s3://my-dvc-study-entry-performance-data
```

### 2. **Ejecutar Pipeline Completo**
```bash
# Descargar datos y ejecutar pipeline
dvc pull
dvc repro

# Verificar resultados
ls -la Equipo54_MLOps/models/
ls -la Equipo54_MLOps/reports/
```

### 3. **Probar Etapas Individuales**
```bash
# Ejecutar solo entrenamiento
dvc repro train

# Forzar re-ejecución completa
dvc repro --force
```

### 4. **Verificar Métricas**
```bash
# Ver métricas del modelo
cat Equipo54_MLOps/reports/metrics/metrics.json

# Ver pipeline visualmente
dvc dag
```

## 📊 Resultados Actuales

El modelo actual genera:
- **Accuracy (Test):** 0.4098
- **Característica más importante:** Caste (73.96%)
- **Distribución de predicciones:** 100% en clase "Excellent"

### Para Nuevos Cambios:
```bash
# Modificar parámetros en params.yaml
dvc repro
git add dvc.lock params.yaml
git commit -m "update: new model parameters"
dvc push
git push
```

### Para Nuevos Datos:
```bash
dvc add nueva_data.csv
git add nueva_data.csv.dvc
git commit -m "feat: add new dataset"
dvc push
git push
```

## 📝 Checklist Antes del Merge

- [x] 🔄 Rama `dev` actualizada (`git pull origin dev`)
- [x] ✅ Pipeline ejecuta completamente sin errores (`dvc repro`)
- [x] 📂 Archivos grandes NO en Git (solo en DVC)
- [x] 🐍 Scripts Python con logging y manejo de errores
- [x] 📊 Métricas generadas y documentadas
- [x] 📚 Documentación actualizada
- [x] 🔧 Configuración DVC + S3 funcional

### 🎯 Puntos de Revisión:
1. **Estructura del pipeline** - Las 5 etapas están correctamente encadenadas
2. **Manejo de dependencias** - DVC detecta cambios automáticamente
3. **Versionado** - Todos los artefactos están versionados via `dvc.lock`
4. **Reproducibilidad** - El pipeline genera resultados consistentes

### 🔮 Próximos Pasos:
- Implementar DVC Experiments para optimización de hiperparámetros
- Agregar validación de datos entre etapas
- Configurar CI/CD con GitHub Actions
- Implementar monitoreo de drift de datos

### ⚠️ Consideraciones:
- El modelo actual muestra sobreajuste (train accuracy 0.5451 vs test 0.4098)
- La característica "Caste" domina las importancias (73.96%)
- Las predicciones actuales están sesgadas hacia una sola clase

**Revisores:** Por favor verificar la reproducibilidad ejecutando `dvc repro` y que la documentación sea clara para nuevos miembros del equipo.

### CODE REVIEW

**Comment by joseluistinajero** [comment]
Aceptado para garantizar ajustes de DVC
